libsolidity/analysis/IntegerOverflowBounds.cpp

```diff
@@ -0,0 +1,267 @@
+	// They fail to cover scenarios like [5,6] < [2,3], which is quite
+	// tautologically false.  Scenarios like that should actually generate
+	// their own warning or something and they could seriously impact
+	// how we can create and apply constraints.
+	//
+	// For now, the LHS passed in here is always the widest range for the type
+	// so this is safe-ish but still wrong.
+	else if (op == Token::LessThan)
+	{
+		max = right.min-1;
+	}
+	else if (op == Token::GreaterThan)
+	{
+		min = right.max+1;
+	}
+	else if (op == Token::Equal)
+	{
+		min = right.min;
+		max = right.max;
+	}
+}
+
+void IntegerOverflowBounds::swapIfNecessary()
+{
+	if (min > max)
+	{
+		dev::bigint swap = min;
+		min = max;
+		max = swap;
+	}
+}
+
+bool IntegerOverflowBounds::fitsType(Type const& type)
+{
+	if (type.category() == Type::Category::Integer)
+		return fitsType(static_cast<IntegerType const&>(type));
+	solAssert(false, "IntegerOverflowBounds fits only IntegerType");
+}
+
+bool IntegerOverflowBounds::fitsType(IntegerType const& type)
+{
+	bool isSigned = type.isSigned();
+	uint16_t bytes = type.storageBytes();
+	uint32_t exponent = bytes * (isSigned ? ((bytes * 8) - 1) : (bytes * 8));
+	dev::bigint allowedMin;
+	dev::bigint allowedMax;
+	if (isSigned)
+	{
+		allowedMin = 1;
+		allowedMin = allowedMin << exponent;
+		allowedMin = -allowedMin;
+
+		allowedMax = 1;
+		allowedMax = allowedMax << exponent;
+		allowedMax = allowedMax - 1;
+	}
+	else
+	{
+		allowedMin = 0;
+		allowedMax = 1;
+		allowedMax = allowedMax << exponent;
+		allowedMax = allowedMax - 1;
+	}
+	return (min >= allowedMin && max <= allowedMax);
+}
+
+void IntegerOverflowBounds::setToMaximumRange(Type const& type)
+{
+	if (type.category() == Type::Category::Integer)
+		setToMaximumRange(static_cast<IntegerType const&>(type));
+	else if (type.category() == Type::Category::RationalNumber)
+		setToMaximumRange(static_cast<RationalNumberType const&>(type));
+	else
+		solAssert(false, string("IntegerOverflowBounds doesn't support type ")+type.toString());
+}
+
+void IntegerOverflowBounds::setToMaximumRange(IntegerType const& type)
+{
+	bool isSigned = type.isSigned();
+	uint16_t bytes = type.storageBytes();
+	uint32_t exponent = bytes * (isSigned ? ((bytes * 8) - 1) : (bytes * 8));
+	dev::bigint allowedMin;
+	dev::bigint allowedMax;
+	if (isSigned)
+	{
+		allowedMin = 1;
+		allowedMin = allowedMin << exponent;
+		allowedMin = -allowedMin;
+
+		allowedMax = 1;
+		allowedMax = allowedMax << exponent;
+		allowedMax = allowedMax - 1;
+	}
+	else
+	{
+		allowedMin = 0;
+		allowedMax = 1;
+		allowedMax = allowedMax << exponent;
+		allowedMax = allowedMax - 1;
+	}
+	min = allowedMin;
+	max = allowedMax;
+}
+
+void IntegerOverflowBounds::setToMaximumRange(RationalNumberType const& type)
+{
+	min = numerator(type.value());
+	max = numerator(type.value());
+}
+
+void IntegerOverflowBounds::setToDefaultValue()
+{
+	min = 0;
+	max = 0;
+}
+
+void IntegerOverflowBounds::constrain(OverflowBounds const& bounds)
```

Comment:
Yes, with the proviso that I'm calling it `intersectBounds` to match `unionBounds`, because `union` is a reserved word.